Sources/Basic/Lock.swift

```diff
@@ -25,3 +26,67 @@ public struct Lock {
         return try body()
     }
 }
+
+enum ProcessLockError: Swift.Error {
+    case unableToAquireLock(errno: Int32)
+}
+
+/// Provides functionality to aquire a lock on a file via POSIX's flock() method.
+/// It can be used for things like serializing concurrent mutations on a shared resource
+/// by mutiple instances of a process.
+public final class FileLock {
+    /// The name of the lock, used in filename of the lock file.
+    let name: String
+
+    /// The directory where the lock file should be created.
+    let cachePath: AbsolutePath
+
+    /// File descriptor to the lock file.
+    private var fd: Int32? = nil
+
+    /// Path to the lock file.
+    private var lockFile: AbsolutePath {
+        return cachePath.appending(component: name + ".lock")
+    }
+
+    /// Create an instance of process lock with a name and the path where
+    /// the lock file can be created.
+    ///
+    /// Note: The cache path should be a valid directory.
+    public init(name: String, cachePath: AbsolutePath) {
+        self.name = name
+        self.cachePath = cachePath
+    }
+
+    /// Try to aquire a lock. This method will block until lock the already aquired by other process.
+    ///
+    /// Note: This method can throw if underlying POSIX methods fail.
+    public func lock() throws {
+        // Open the lock file.
+        let fp = libc.fopen(lockFile.asString, "w")
+        if fp == nil {
+            throw FileSystemError(errno: errno)
+        }
+        // Save the fd to close and remove lock later.
+        fd = fileno(fp)
+        // Aquire lock on the file.
+        guard flock(fd!, LOCK_EX) == 0 else {
```

Comment:
As discussed on slack, there is no good solution to file locking which works with NFS too so we're going to start with `flock()` and see how it goes. The underlying mechanism can be changed anytime and the current usecase is just a protective measure for now.